### DISCUSSION TITLE
[Issue][Native] All icons are added to the native bundle

### DISCUSSION THREAD

**Discussion by matteo-hertel** [tamagui/tamagui#1634] [General]
### Current Behavior

When importing any component from Tamagui in a native app, all of the icons are added to the bundle, the screenshot below is from the kitchen sink app in this repo, there are 4 icons used but all of them are added to the bundle

![Screenshot 2023-08-09 at 12 12 09 PM](https://github.com/tamagui/tamagui/assets/3579758/7f2dd98a-dc74-4d1a-922d-87e9c68ed1a9)

### Expected Behavior

Only used icons directly or indirectly should be added to the bundle
Worth noting that Metro doesn't directly support tree shaking (see https://github.com/facebook/metro/issues/227) but there is definitely something that can be done to prevent all the icons to be imported when importing anything from Tamagui

while investigating this on my app, I removed everything except importing a single`YStack` and all icons were added to the bundle

### Tamagui Version

```markdown
v1.52.0
Tested on `9353e926f1e5d6b1001c052267359bca789e5ff9`
```

### Reproduction

```markdown
I haven't forked the repo as there isn't much to except run `react-native-bundle-visualize` on
https://github.com/tamagui/tamagui/tree/master/apps/kitchen-sink

Step by step:

- clone the repo or switch to master
- `yarn && yarn build` so that all the packages are build
- cd `apps/kitchen-sink`
- `yarn add -D react-native-bundle-visualize` this is needed as this is a monorepo, optional for single repos
- run `npx react-native-bundle-visualizer --only-mapped --bundle-output ./myapp.bundle --format html --reset-cache --expo true`
- it will open a page in your browser with a bundle explorer and all the icons will show up in the bundle
```

Any further question or context please ask, hopefully I painted the right picture

**Comment by natew** [comment]
You'd have to import them directly, or we'd need a babel plugin that auto-rewrites the imports which is a good idea. Converting to discussion since its more of a feature request.

**Reply by isaachinman** [reply]
@natew Has there been any progress on this? Importing directly sounds fine, but we'd need resolutions via package.json and TS types.

**Reply by isaachinman** [reply]
@natew For the sake of speed, I had to resort to copy/pasting the source code into my vc. This took ~1.5MB off my bundle. A first-class solution to this would be highly appreciated 🙏

**Comment by vonovak** [comment]
you can use https://microsoft.github.io/rnx-kit/docs/guides/bundling#tree-shaking

it basically runs the metro-produced code through esbuild, afaik

**Reply by isaachinman** [reply]
Hi @vonovak, do you have an actual example or minimal repro of using this with Tamagui+Expo?

**Comment by wcastand** [comment]
for those coming here from expo, rnx-kit is a bit annoying (if even possible) to match with expo easyly so i tried something else. and use babel transform on lucide-icon.

https://www.npmjs.com/package/babel-plugin-transform-imports


```js
// config.babel.js
const wordSeparators = /[\s\u2000-\u206F\u2E00-\u2E7F\\'!"#$%&()*+,\-.\/:;<=>?@\[\]^_`{|}~]+/
const capital_plus_lower = /[A-ZÀ-Ý\u00C0-\u00D6\u00D9-\u00DD][a-zà-ÿ]/g
const capitals = /[A-Z0-9À-Ý\u00C0-\u00D6\u00D9-\u00DD]+/g

function kebabCase(str) {
	let res = str
	// replace word starts with space + lower case equivalent for later parsing
	// 1) treat cap + lower as start of new word
	res = str.replace(
		capital_plus_lower,
		(match) =>
			// match is one caps followed by one non-cap
			` ${match[0].toLowerCase() || match[0]}${match[1]}`,
	)
	// 2) treat all remaining capitals as words
	res = res.replace(
		capitals,
		(match) =>
			// match is a series of caps
			` ${match.toLowerCase()}`,
	)
	return res.trim().split(wordSeparators).join("-").replace(/^-/, "").replace(/-\s*$/, "")
}

module.exports = (api) => {
	api.cache(true)
	return {
		presets: [["babel-preset-expo", { jsxRuntime: "automatic" }]],
		env: {
			production: {
				plugins: ["transform-remove-console"],
			},
		},
		plugins: [
			[
				"transform-imports",
				{
					"@tamagui/lucide-icons": {
						transform: (importName) => `@tamagui/lucide-icons/dist/esm/icons/${kebabCase(importName)}`,
						preventFullImport: true,
						skipDefaultConversion: true,
					},
				},
			],
			"react-native-reanimated/plugin",
		],
	}
}

```

**Reply by matteo-hertel** [reply]
I finally had the time to look into this.
Thank you so much for providing the babel-transform!

I just dropped this in and I had 34.7 % reduction in bundle size.
From 21.9 MB to 14.3MB

Thank you so much!

**Reply by wcastand** [reply]
worth noting, it's only on the lucide icons, so the whole tamagui is still imported outside of icons even if you only import one `YStack`

it's not full tree shaking of tamagui in case someone find this too :) unless you import from tamagui one by one like explain earlier in the comments.

still need to get metro to do tree shaking to really keep the size down to only what you use. (In the case of tamagui, i'd say it won't win much outside of icons tho since you probably use most of it)

**Reply by matteo-hertel** [reply]
I kept an eye on my bundle size and the icons were the one thing that bugged me. 
I wish I had thought about using a transformer.

Using expo atlas this is before adding your transformer

![Screenshot 2024-06-25 at 15 36 29](https://github.com/tamagui/tamagui/assets/3579758/6d23fb95-493d-42d5-9a2d-d5a30d909762)

and this is after

![Screenshot 2024-06-25 at 15 37 17](https://github.com/tamagui/tamagui/assets/3579758/6bf0da4d-7a4c-435c-b0d1-b81d65bdccb8)

the biggest module in the entire codebase is gone

so for me was a smashing result.
Now I have to go fix my lodash imports 😇

**Reply by isaachinman** [reply]
This is unfortunately not working for me with Expo v53.0.20. Any new approaches?

**Reply by ghaschel** [reply]
@isaachinman 

Just change this:
> ```js
> @tamagui/lucide-icons/dist/esm/icons/${kebabCase(importName)}`,
> ```

to this:
> ```js
> @tamagui/lucide-icons/dist/esm/icons/${importName}.native.js`,
> ```

**Reply by isaachinman** [reply]
I believe this is completely resolved in v2.

**Reply by ghaschel** [reply]
I'm still facing this in Tamagui `2.0.0-rc.23` and Expo `55.0.5`

**Reply by isaachinman** [reply]
Ah, sorry to hear it. I've stopped using Lucide so I no longer have any advice or input. You may want to raise this in Discord.
